crates/secret-service-proto/src/v1/traits.rs

```diff
@@ -0,0 +1,164 @@
+use std::future::Future;
+
+use bitcoin::Txid;
+use musig2::{
+    errors::{RoundContributionError, RoundFinalizeError},
+    secp256k1::{schnorr::Signature, PublicKey},
+    AggNonce, LiftedSignature, PartialSignature, PubNonce,
+};
+use quinn::{ConnectionError, ReadExactError, WriteError};
+use rkyv::{rancor, Archive, Deserialize, Serialize};
+use strata_bridge_primitives::scripts::taproot::TaprootWitness;
+
+use super::wire::ServerMessage;
+
+pub trait SecretServiceFactory<FirstRound, SecondRound>: Send + Clone
+where
+    FirstRound: Musig2SignerFirstRound<Server, SecondRound> + Clone,
+    SecondRound: Musig2SignerSecondRound<Server>,
+{
+    type Context: Send + Clone;
+    type Service: SecretService<Server, FirstRound, SecondRound> + Send;
+    fn produce(ctx: Self::Context) -> Self::Service;
+}
+
+// possible when https://github.com/rust-lang/rust/issues/63063 is stabliized
+// pub type AsyncResult<T, E = ()> = impl Future<Output = Result<T, E>>;
+
+pub trait SecretService<O, FirstRound, SecondRound>: Send
+where
+    O: Origin,
+    FirstRound: Musig2SignerFirstRound<O, SecondRound>,
+{
+    type OperatorSigner: OperatorSigner<O>;
+    type P2PSigner: P2PSigner<O>;
+    type Musig2Signer: Musig2Signer<O, FirstRound>;
+    type WotsSigner: WotsSigner<O>;
+    type StakeChain: StakeChainPreimages<O>;
+
+    fn operator_signer(&self) -> Self::OperatorSigner;
+    fn p2p_signer(&self) -> Self::P2PSigner;
+    fn musig2_signer(&self) -> Self::Musig2Signer;
+    fn wots_signer(&self) -> Self::WotsSigner;
+    fn stake_chain(&self) -> Self::StakeChain;
```

Comment:
I got the idea but this is **_not_** the Stake Chain.
This is the preimages for a Stake transaction in the chain.
We need better naming.

Reply:
Maybe 
```suggestion
    fn stake_preimages(&self) -> Self::StakePreimages;
```